File: solutions/java/0057.java
/*
 * Leetcode 57: Insert Interval. Java. Medium.
 * O(n) runtime and O(n) space complexity.
 * This solution processes the input intervals by iterating through them once, handling three cases:
 * intervals before the new interval, intervals after the new interval, and overlapping intervals.
 * It merges overlapping intervals on the fly and adds the new interval when appropriate, ensuring optimal performance.
 * Insights: The direct traversal approach leverages the sorted nature of intervals for simplicity and efficiency.
 * 12/26/2024 Winston Tsui
 */

import java.util.*;

class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        ArrayList<int[]> ans = new ArrayList<>();

        for (int i = 0; i < intervals.length; i++) {
            // If this interval goes before newInterval
            if (intervals[i][1] < newInterval[0])
                ans.add(intervals[i]);

            // If this interval goes after newInterval
            else if (intervals[i][0] > newInterval[1]) {
                ans.add(newInterval);

                // Copy everything else into ans
                while (i < intervals.length)
                    ans.add(intervals[i++]);
                return ans.toArray(new int[ans.size()][]);
            }

            // If there is an insertion merge conflict
            else {
                newInterval[0] = Math.min(intervals[i][0], newInterval[0]);
                newInterval[1] = Math.max(intervals[i][1], newInterval[1]);
            }
        }
        ans.add(newInterval);
        return ans.toArray(new int[ans.size()][]);
    }
}

File: solutions/python/0057.py
"""
 * Leetcode 57: Insert Interval. Python. Medium.
 * O(N) runtime, O(N) space complexity
 * One-pass approach to handle three cases: before, after, and overlap with new interval.
 *
 * Example: intervals = [[1,3],[6,9]], newInterval = [2,5]
 * Step 1: [1,3] overlaps with [2,5] -> merge to [1,5]
 * Step 2: [6,9] comes after [1,5] -> add to result
 * Result: [[1,5],[6,9]]
 *
 * 7/5/2025 Winston Tsui
"""

from typing import List

class Solution:
    def insert(
        self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        inserted = []

        for i in range(len(intervals)):
            # New interval comes before this
            if newInterval[1] < intervals[i][0]:
                inserted.append(newInterval)
                return inserted + intervals[i:]
            # New interval comes after this
            elif newInterval[0] > intervals[i][1]:
                inserted.append(intervals[i])
            # There's a merge conflict
            else:
                newInterval = [
                    min(newInterval[0], intervals[i][0]),
                    max(newInterval[1], intervals[i][1]),
                ]

        inserted.append(newInterval)

        return inserted
